components/scream/src/control/surface_coupling.cpp

```diff
@@ -316,21 +316,23 @@ void SurfaceCoupling::do_export (const bool init_phase)
       const auto T_mid_i          = ekat::subview(T_mid, i);
       const auto p_mid_i          = ekat::subview(p_mid, i);
       const auto pseudo_density_i = ekat::subview(pseudo_density, i);
+      const auto dz_i             = ekat::subview(dz, i);
+      const auto z_int_i          = ekat::subview(z_int, i);
+      const auto z_mid_i          = ekat::subview(z_mid, i);
 
-      // Compute vertical layer heights
-      Kokkos::parallel_for(Kokkos::TeamThreadRange(team, m_num_levs), [&] (const Int& k) {
-        dz(k) = PF::calculate_dz(pseudo_density_i(k), p_mid_i(k), T_mid_i(k), qv_i(k));
-      });
+      // Compute vertical layer thickness
+      PF::calculate_dz(team, pseudo_density_i, p_mid_i, T_mid_i, qv_i, dz_i);
       team.team_barrier();
 
-      // Compute vertical layer heights
-      PF::calculate_z_int(team,m_num_levs,dz,0.0,z_int);
+      // Compute vertical layer heights. Use z_int(nlevs) = z_surf = 0.0.
+      const Real z_surf = 0.0;
+      PF::calculate_z_int(team, m_num_levs, dz_i, z_surf, z_int_i);
```

Comment:
Since you are using `KOKKOS_LAMBDA`, you can't access class members from inside. Here, `m_num_levs` causes a `this->m_num_levs` access on device. You need a local copy before the parallel for.